File: apps/agent/packages/tui/src/acp/skills.ts
import * as acp from '@agentclientprotocol/sdk';

import type { TuiSkill } from '../types/runtime-models.js';
import type { TuiAcpRuntime } from './runtime.js';

/// One Skill, as an ACP client is told about it.
///
/// **The agent's own row, unchanged.** A client draws a name, a sentence and a
/// switch, and all three are facts this runtime already holds — `enabled` included,
/// since a Skill switched off is one the catalog no longer contributes to the
/// model. `locationUri` rides along because the registry keys its disabled set by
/// it: a client that hands back only a *name* sets nothing the moment the Skill is
/// renamed or re-seeded, and it would fail silently.
export interface TuiAcpSkill {
  readonly name: string;
  readonly displayName: string;
  readonly description: string;
  readonly enabled: boolean;
  readonly sourceKind?: string;
  readonly locationUri?: string;
}

/// The Skill roster, and the two things a client may do to one.
///
/// **These exist because the terminal's roster was not reachable from a client.**
/// The TUI lists Skills for its own composer and switches them from its own panel,
/// and an ACP client — which has no terminal — had `/skills` and nothing else: a
/// list in prose it would have had to parse, and no way to read or switch one. The
/// runtime already holds all three, so this is the seam rather than a second
/// implementation.
///
/// **The roster is the management view, not the runtime one.** `listSkills` answers
/// what the *model* is told about, and a Skill switched off is filtered out of it —
/// so a client built on that could switch a Skill off and then never see it again
/// to switch it back on. [`listAllSkills`](TuiAcpRuntime) keeps every Skill and
/// marks each one, which is the view a screen with switches in it needs.
export function registerTuiAcpSkillExtensions(options: {
  readonly app: acp.AgentApp;
  readonly runtime: TuiAcpRuntime;
  readonly resolveSession: (
    sessionId: string,
  ) => { readonly agentName?: string } | undefined;
}): void {
  const agentNameFor = (sessionId: string): string | undefined =>
    options.resolveSession(sessionId)?.agentName;

  options.app.onRequest('mcode/session/skills/list', parseListRequest, async ({ params }) => {
    const result = await options.runtime.listAllSkills(agentNameFor(params.sessionId));
    return {
      skills: (result?.skills ?? []).map(toAcpSkill),
      // Rides the answer rather than being assumed away: a client that drew a
      // count would otherwise be stating something it was never told.
      hasMore: result?.hasMore === true,
    };
  });

  options.app.onRequest('mcode/session/skills/read', parseReadRequest, async ({ params }) => ({
    name: params.name,
    // **`null`, not a throw.** A Skill deleted between the list and the press is
    // an answer a client can draw — "this one is gone" — where an error only says
    // that something went wrong, which is the wrong thing to put under a heading
    // that already named the Skill.
    text: (await options.runtime.readSkill(params.name, params.locationUri)) ?? null,
  }));

  options.app.onRequest(
    'mcode/session/skills/set_enabled',
    parseSetEnabledRequest,
    async ({ params }) => {
      // **`applied` is not `enabled`.** The registry answers nothing for a Skill it
      // does not recognise, and that is the honest reply rather than a throw — the
      // row the client pressed may have been deleted since it was drawn. A client
      // that read the flag for the new state would draw a switch that moved over a
      // Skill nobody changed.
      const applied = await options.runtime.setSkillEnabled(
        params.name,
        params.enabled,
        params.locationUri,
      );
      return { name: params.name, enabled: params.enabled, applied };
    },
  );
}

function toAcpSkill(skill: TuiSkill): TuiAcpSkill {
  const name = skill.name.trim();
  return {
    name,
    displayName: skill.displayName?.trim() || name,
    description: (skill.displayDescription ?? skill.description ?? '').trim(),
    // Absent reads as on, which is what the registry's own default is: a Skill
    // nobody has switched off is one the model is told about.
    enabled: skill.enabled !== false,
    ...(skill.sourceKind ? { sourceKind: skill.sourceKind } : {}),
    ...(skill.locationUri ? { locationUri: skill.locationUri } : {}),
  };
}

function parseListRequest(value: unknown): { sessionId: string } {
  const record = requireRecord(value);
  return { sessionId: requireText(record.sessionId, 'sessionId') };
}

function parseReadRequest(value: unknown): SkillTarget {
  return parseSkillTarget(value);
}

function parseSetEnabledRequest(value: unknown): SkillTarget & { enabled: boolean } {
  const enabled = requireRecord(value).enabled;
  if (typeof enabled !== 'boolean') {
    throw acp.RequestError.invalidParams(undefined, '`enabled` must be a boolean.');
  }
  return { ...parseSkillTarget(value), enabled };
}

interface SkillTarget {
  readonly sessionId: string;
  readonly name: string;
  readonly locationUri?: string;
}

/// Everything both Skill requests carry: which Session, which Skill, and the
/// location the row was drawn from.
function parseSkillTarget(value: unknown): SkillTarget {
  const record = requireRecord(value);
  return {
    sessionId: requireText(record.sessionId, 'sessionId'),
    name: requireText(record.name, 'name'),
    ...(record.locationUri === undefined
      ? {}
      : { locationUri: requireText(record.locationUri, 'locationUri') }),
  };
}

function requireRecord(value: unknown): Record<string, unknown> {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) {
    throw acp.RequestError.invalidParams(undefined, 'Expected an object.');
  }
  return value as Record<string, unknown>;
}

function requireText(value: unknown, field: string): string {
  if (typeof value !== 'string' || value.trim().length === 0) {
    throw acp.RequestError.invalidParams(undefined, `\`${field}\` must be a non-empty string.`);
  }
  return value.trim();
}

File: apps/desktop/src-tauri/src/harness/mcode/skills.rs
//! What the agent answers about the reader's Skills.
//!
//! **The screen that draws this is not a session and does not own one.** Every
//! call here takes the session it is asked on — the control child
//! [`open_control`](super::open_control) keeps — and holds nothing, so that
//! child's lifecycle stays in one place rather than being smeared across the
//! readers of it.
//!
//! **The roster read here is the management one, not the runtime one.** The agent
//! has two lists: the runtime list is what the *model* is told about, and a Skill
//! switched off is filtered out of it; this is the one the app's own "My skills"
//! reads, which keeps every row and marks it. Building a screen with switches on
//! the first list is how a Skill gets switched off and then never seen again —
//! there would be no row left to switch it back on.
//!
//! Rejected: reading `SKILL.md` here and handing the frontend the text directly.
//! The registry owns where a Skill lives and what its body is, and a path parsed
//! out of a `files://` URI is this app keeping a copy of a fact that moves.

use anyhow::{Context, Result};
use serde_json::{json, Value};

use super::McodeSession;
use crate::plugins::{PluginSkill, SkillRoster, SkillToggle};

/// The whole roster, Skills switched off included and marked.
pub async fn roster(session: &McodeSession) -> Result<SkillRoster> {
    let reply = session
        .client
        .request("mcode/session/skills/list", json!({"sessionId": session.id}))
        .await
        .context("the agent refused to list its Skills")?;

    Ok(SkillRoster {
        skills: reply
            .get("skills")
            .and_then(Value::as_array)
            .map(|rows| rows.iter().filter_map(skill_of).collect())
            .unwrap_or_default(),
        // Rides the answer rather than being assumed away: a screen drawing a
        // count would otherwise be stating something it was never told.
        has_more: reply
            .get("hasMore")
            .and_then(Value::as_bool)
            .unwrap_or(false),
    })
}

/// Switches one Skill, and answers whether the registry took it.
///
/// **A refusal is not an error here.** The registry answers nothing for a Skill it
/// does not recognise — a row deleted or renamed since it was drawn — and that
/// arrives as `applied: false` on a successful call, which is the sentence the
/// screen can draw. What an `Err` means is the transport, so the caller drops the
/// child rather than retrying into a refusal.
pub async fn set_enabled(
    session: &McodeSession,
    name: &str,
    enabled: bool,
    location_uri: Option<&str>,
) -> Result<SkillToggle> {
    let reply = session
        .client
        .request(
            "mcode/session/skills/set_enabled",
            json!({
                "sessionId": session.id,
                "name": name,
                "enabled": enabled,
                // The registry keys its disabled set by this, so a row built from
                // a location hands it back: a rename between the read and the
                // press would otherwise leave the switch setting nothing, and it
                // would fail silently.
                "locationUri": location_uri,
            }),
        )
        .await
        .context("the agent refused to switch this Skill")?;

    Ok(SkillToggle {
        name: reply
            .get("name")
            .and_then(Value::as_str)
            .unwrap_or(name)
            .to_string(),
        enabled: reply
            .get("enabled")
            .and_then(Value::as_bool)
            .unwrap_or(enabled),
        // Absent reads as *not* applied: a screen that claimed the switch took
        // when the agent said nothing would move a control over nothing.
        applied: reply
            .get("applied")
            .and_then(Value::as_bool)
            .unwrap_or(false),
    })
}

/// One Skill's own text, or `None` for one the registry cannot find.
pub async fn read(
    session: &McodeSession,
    name: &str,
    location_uri: Option<&str>,
) -> Result<Option<String>> {
    let reply = session
        .client
        .request(
            "mcode/session/skills/read",
            json!({"sessionId": session.id, "name": name, "locationUri": location_uri}),
        )
        .await
        .context("the agent refused to read this Skill")?;

    // `null` is the agent's word for a Skill that is gone. It stays `None` rather
    // than becoming an error: the row the reader pressed may have been deleted
    // since it was listed, and "this one is gone" is a sentence, not a failure.
    Ok(reply
        .get("text")
        .and_then(Value::as_str)
        .map(str::to_string))
}

/// One row of the roster, out of the agent's own object.
///
/// A row this build cannot name is dropped rather than drawn as an empty line —
/// nothing on screen can be done with it. `enabled` is the exception and the
/// reason this is not a strict parse: absent means *on*, which is the registry's
/// own default, and reading it the other way would draw every Skill as switched
/// off on a reply that predates the field.
fn skill_of(row: &Value) -> Option<PluginSkill> {
    let name = text(row, "name")?;
    Some(PluginSkill {
        display_name: text(row, "displayName").unwrap_or_else(|| name.clone()),
        name,
        description: text(row, "description").unwrap_or_default(),
        enabled: row
            .get("enabled")
            .and_then(Value::as_bool)
            .unwrap_or(true),
        source_kind: text(row, "sourceKind"),
        location_uri: text(row, "locationUri"),
    })
}

/// A trimmed non-empty string, or `None`.
fn text(row: &Value, key: &str) -> Option<String> {
    let value = row.get(key).and_then(Value::as_str)?.trim();
    (!value.is_empty()).then(|| value.to_string())
}
